--- qemu-10.1.2/include/ui/sdl2.h2025-10-20 15:44:00
+++ qemu-10.1.2-patched/include/ui/sdl2.h2025-11-26 09:07:45
@@ -45,6 +48,9 @@
     bool gui_keysym;
     SDL_GLContext winctx;
     QKbdState *kbd;
+#ifdef CONFIG_SDL_CLIPBOARD
+    QemuClipboardPeer cbpeer;
+#endif
 #ifdef CONFIG_OPENGL
--- qemu-10.1.2/ui/sdl2.c2025-10-20 15:44:01
+++ qemu-10.1.2-patched/ui/sdl2.c2025-11-26 09:07:45
@@ -616,15 +616,21 @@
          * key is released.
          */
         scon->ignore_hotkeys = get_mod_state();
+#ifdef CONFIG_SDL_CLIPBOARD
+        sdl2_clipboard_handle_focus_change(scon, true);
+#endif
         break;
     case SDL_WINDOWEVENT_FOCUS_LOST:
         if (!fxui_focus_lost() && gui_grab && !gui_fullscreen) {
             fxui_grab_val(0x80 | gui_grab);
             sdl_grab_end(scon);
         }
+#ifdef CONFIG_SDL_CLIPBOARD
+        sdl2_clipboard_handle_focus_change(scon, false);
+#endif
         break;
diff --git a/ui/vdagent.c b/ui/vdagent.c
index bb0c4aa..6369cd8 100644
--- a/ui/vdagent.c
+++ b/ui/vdagent.c
@@ -197,7 +197,8 @@ static void vdagent_send_caps(VDAgentChardev *vd, bool request)
         caps->caps[0] |= (1 << VD_AGENT_CAP_CLIPBOARD_BY_DEMAND);
         caps->caps[0] |= (1 << VD_AGENT_CAP_CLIPBOARD_SELECTION);
 #if CHECK_SPICE_PROTOCOL_VERSION(0, 14, 1)
         caps->caps[0] |= (1 << VD_AGENT_CAP_CLIPBOARD_GRAB_SERIAL);
+        caps->caps[0] |= (1 << VD_AGENT_CAP_CLIPBOARD_NO_RELEASE_ON_REGRAB);
 #endif
     }