--- a/ui/sdl2.c
+++ b/ui/sdl2.c
@@ -99,5 +99,14 @@ void sdl2_window_create(struct sdl2_console *scon)
         if (scon->opts->gl == DISPLAY_GL_MODE_ES) {
             driver = "opengles2";
+            SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK,
+                                SDL_GL_CONTEXT_PROFILE_ES);
+#ifdef __MINGW32__
+            SDL_SetHint(SDL_HINT_OPENGL_ES_DRIVER, "1");
+            SDL_SetHint("SDL_ANGLE_BACKEND", "d3d11");
+            SDL_SetHint("SDL_ANGLE_FAST_PATH", "1");
+            SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 3);
+            SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 0);
+#endif
         }
 
         SDL_SetHint(SDL_HINT_RENDER_DRIVER, driver);
