diff --git a/accel/stubs/whpx-stub.c b/accel/stubs/whpx-stub.c
index 4529dc4..a05e738 100644
--- a/accel/stubs/whpx-stub.c
+++ b/accel/stubs/whpx-stub.c
@@ -11,3 +11,7 @@
 
 bool whpx_allowed;
 bool whpx_irqchip_in_kernel;
+void whpx_update_guest_pa_range(uint64_t start_pa, uint64_t size,
+                                void *host_va, int readonly, int add);
+void whpx_update_guest_pa_range(uint64_t start_pa, uint64_t size,
+                                void *host_va, int readonly, int add) {}
diff --git a/accel/whpx/whpx-common.c b/accel/whpx/whpx-common.c
index b813a5d..a899cd5 100644
--- a/accel/whpx/whpx-common.c
+++ b/accel/whpx/whpx-common.c
@@ -13,6 +13,7 @@
 #include "qemu/accel.h"
 #include "accel/accel-ops.h"
 #include "system/memory.h"
+#include "system/ramblock.h"
 #include "system/whpx.h"
 #include "system/cpus.h"
 #include "system/runstate.h"
@@ -332,6 +333,40 @@ static void whpx_set_phys_mem(MemoryRegionSection *section, bool add)
     }
 }
 
+void whpx_update_guest_pa_range(uint64_t start_pa, uint64_t size, void *host_va,
+                                int readonly, int add)
+{
+    MemoryRegion mr;
+    MemoryRegionSection section;
+    RAMBlock ram_block;
+
+    memset(&ram_block, 0, sizeof(RAMBlock));
+    ram_block.mr = &mr;
+    ram_block.used_length = REAL_HOST_PAGE_ALIGN(size);
+    ram_block.max_length = REAL_HOST_PAGE_ALIGN(size);
+    ram_block.fd = -1;
+    ram_block.guest_memfd = -1;
+    ram_block.page_size = qemu_real_host_page_size();
+    ram_block.host = host_va;
+    ram_block.flags |= RAM_PREALLOC;
+
+    memory_region_init(&mr, NULL, NULL, REAL_HOST_PAGE_ALIGN(size));
+    mr.ram = true;
+    mr.ram_block = &ram_block;
+    mr.readonly = readonly;
+    mr.nonvolatile = false;
+
+    memset(&section, 0, sizeof(section));
+    section.mr = &mr;
+    section.offset_within_region = 0;
+    section.size = int128_make64(REAL_HOST_PAGE_ALIGN(size));
+    section.offset_within_address_space = start_pa;
+    section.readonly = mr.readonly;
+    section.nonvolatile = mr.nonvolatile;
+
+    whpx_set_phys_mem(&section, add);
+    object_unref(OBJECT(&mr));
+}
 static void whpx_region_add(MemoryListener *listener,
                            MemoryRegionSection *section)
 {
diff --git a/meson.build b/meson.build
index d3e4ea0..17da769 100644
--- a/meson.build
+++ b/meson.build
@@ -302,7 +302,7 @@ elif cpu == 'x86_64'
   accelerator_targets += {
     'CONFIG_HVF': ['x86_64-softmmu'],
     'CONFIG_NITRO': ['x86_64-softmmu'],
     'CONFIG_NVMM': ['i386-softmmu', 'x86_64-softmmu'],
-    'CONFIG_WHPX': ['x86_64-softmmu'],
+    'CONFIG_WHPX': ['i386-softmmu', 'x86_64-softmmu'],
     'CONFIG_MSHV': ['x86_64-softmmu'],
   }
