diff --git a/target/i386/whpx/whpx-all.c b/target/i386/whpx/whpx-all.c
index e56ae2b..e98c833 100644
--- a/target/i386/whpx/whpx-all.c
+++ b/target/i386/whpx/whpx-all.c
@@ -1960,18 +1960,23 @@ int whpx_vcpu_run(CPUState *cpu)
                 cpu->exception_index = EXCP_INTERRUPT;
             } else {
                 /* Another exception or debug event. Report it to GDB. */
                 cpu->exception_index = EXCP_DEBUG;
             }
 
             ret = 1;
             break;
-        case WHvRunVpExitReasonNone:
         case WHvRunVpExitReasonUnrecoverableException:
+            whpx_get_registers(cpu, WHPX_LEVEL_FULL_STATE);
+            error_report("WHPX: unrecoverable guest exception at rip=0x%" PRIx64,
+                         (uint64_t)cpu_env(cpu)->eip);
+            ret = -1;
+            break;
+        case WHvRunVpExitReasonNone:
         case WHvRunVpExitReasonInvalidVpRegisterValue:
         case WHvRunVpExitReasonUnsupportedFeature:
         default:
             error_report("WHPX: Unexpected VP exit code %d",
                          vcpu->exit_ctx.ExitReason);
             whpx_get_registers(cpu, WHPX_LEVEL_FULL_STATE);
             bql_lock();
             qemu_system_guest_panicked(cpu_get_crash_info(cpu));
